### PR TITLE
Harrison/docs combine document chain

### PR DESCRIPTION
**Pull request by hwchase17**

### CODE REVIEW

File: langchain/chains/combine_documents/reduce.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    We pass all previous results to this chain, and the output of this chain is
+    returned as a final result.
+
+    `collapse_document_chain` is used if the documents passed in are too many to all
+    be passed to `combine_document_chain` in one go. In this case,
+    `collapse_document_chain` is called recursively on as big of groups of documents
+    as are allowed.
+    """
+
+    combine_document_chain: BaseCombineDocumentsChain
```
Comment (baskaryan) [inline langchain/chains/combine_documents/reduce.py:150]: while we're refactoring would vote to rename combine_document*s*_chain

File: langchain/schema.py [hunk 2]
```diff
@@ -646,3 +646,49 @@ def save(self, file_path: Union[Path, str]) -> None:
...
+        .. code-block:: python
+        from langchain.schema import Document
+        from langchain.prompts import PromptTemplate
+        doc = Document(page_content="This is a joke", metadata={"page": "1"})
+        prompt = PromptTemplate.from_template("Page {page}: {page_content}")
+        format_document(doc, prompt)
+        >>> "Page 1: This is a joke"
+    """
+    base_info = {"page_content": doc.page_content}
+    base_info.update(doc.metadata)
```
Comment (baskaryan) [inline langchain/schema.py:682]: could we just include `**doc.metadata` in dict literal?

File: langchain/schema.py [hunk 3]
```diff
@@ -646,3 +646,49 @@ def save(self, file_path: Union[Path, str]) -> None:
...
+    missing_metadata = set(prompt.input_variables).difference(base_info)
```
Comment (baskaryan) [inline langchain/schema.py:683]: ooc will prompt template not already handle checking this

File: langchain/schema.py [hunk 4]
```diff
@@ -646,3 +646,49 @@ def save(self, file_path: Union[Path, str]) -> None:
                 yaml.dump(prompt_dict, f, default_flow_style=False)
         else:
             raise ValueError(f"{save_path} must be json or yaml")
+
+
+def format_document(doc: Document, prompt: BasePromptTemplate) -> str:
```
Comment (baskaryan) [inline langchain/schema.py:651]: for sep pr but what if we had
```python
class DocumentPromptTemplate(PromptTemplate):
  doc_content_key: str = "page_content"
  def format(self, document: Optional[Document]=None, **kwargs: Any) -> str:
    if document:
      kwargs[self.doc_content_key] = doc.page_content
      kwargs.update(doc.metadata)
    return super().format(**kwargs)
```

and then you could easily make LLMChains that take docs

File: langchain/chains/combine_documents/map_reduce.py [hunk 5]
```diff
@@ -2,74 +2,39 @@
...
+    """
 
     llm_chain: LLMChain
     """Chain to apply to each document individually."""
-    combine_document_chain: BaseCombineDocumentsChain
-    """Chain to use to combine results of applying llm_chain to documents."""
-    collapse_document_chain: Optional[BaseCombineDocumentsChain] = None
-    """Chain to use to collapse intermediary results if needed.
-    If None, will use the combine_document_chain."""
+    reduce_document_chain: BaseCombineDocumentsChain
```
[the remark this answers is not in the document]
Comment (hwchase17) [reply langchain/chains/combine_documents/map_reduce.py:89]: i think could also be StuffDocumentsChain

File: langchain/schema.py (hunk 3 above)
Comment (hwchase17) [reply langchain/schema.py:683]: nope

File: langchain/vectorstores/cassandra.py [hunk 6]
```diff
@@ -237,10 +238,12 @@ def similarity_search(
         k: int = 4,
         **kwargs: Any,
     ) -> List[Document]:
+        #
```
Comment (nfcampos) [inline langchain/vectorstores/cassandra.py:241]: what's this for?

File: langchain/vectorstores/cassandra.py [hunk 7]
```diff
@@ -125,39 +131,33 @@ def add_texts(
...
+            _texts, embedding_vectors, ids, metadatas
+        ):
+            self.table.put(
+                document=text,
+                embedding_vector=embedding_vector,
+                document_id=text_id,
+                metadata=metadata,
+                ttl_seconds=ttl_seconds,
+            )
         #
```
Comment (nfcampos) [inline langchain/vectorstores/cassandra.py:160]: is this here on purpose?

File: langchain/chains/combine_documents/reduce.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+
+    This involves
+    - combine_document_chain
+    - collapse_document_chain
+
+    `combine_document_chain` is ALWAYS provided. This is final chain that is called.
+    We pass all previous results to this chain, and the output of this chain is
+    returned as a final result.
+
+    `collapse_document_chain` is used if the documents passed in are too many to all
```
Comment (baskaryan) [inline langchain/chains/combine_documents/reduce.py:91]: nit: `collapse_document_chain` is used if there are more documents than can be passed to `combine_document_chain` in a single call.

File: langchain/chains/combine_documents/reduce.py [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+                input_documents=docs, callbacks=callbacks, **kwargs
+            )
+
+        while num_tokens is not None and num_tokens > token_max:
+            new_result_doc_list = _split_list_of_docs(
+                result_docs, length_func, token_max, **kwargs
+            )
+            result_docs = []
+            for docs in new_result_doc_list:
+                new_doc = _collapse_docs(docs, _collapse_docs_func, **kwargs)
```
Comment (baskaryan) [inline langchain/chains/combine_documents/reduce.py:242]: why can't we call _collapse_docs_func first and pass output to _collapse_docs (and then we don't need to define _collapse_docs_func in first place)

File: langchain/vectorstores/cassandra.py
Comment (baskaryan) [inline langchain/vectorstores/cassandra.py:1]: guessing this whole file of changes is unintentional?

File: langchain/vectorstores/cassandra.py [hunk 10]
```diff
@@ -51,7 +59,7 @@ def __init__(
         session: Session,
         keyspace: str,
         table_name: str,
-        ttl_seconds: Optional[int] = None,
+        ttl_seconds: int | None = CASSANDRA_VECTORSTORE_DEFAULT_TTL_SECONDS,
```
Comment (baskaryan) [inline langchain/vectorstores/cassandra.py:62]: | is no bueno

**Pull request merged**